FailedBonusActivationActivity/handler.ts

```diff
@@ -0,0 +1,197 @@
+  bonusLeaseModel: BonusLeaseModel,
+  familyUID: FamilyUID
+): TaskEither<Error, void> => {
+  return fromQueryEither(() => bonusLeaseModel.deleteOneById(familyUID)).map(
+    _ => void 0
+  );
+};
+
+const updateBonusAsFailed = (
+  bonusActivationModel: BonusActivationModel,
+  bonusActivation: BonusActivationWithFamilyUID
+): TaskEither<Error, RetrievedBonusActivation> => {
+  return fromQueryEither(() => {
+    const bonusToUpdate: NewBonusActivation = {
+      ...bonusActivation,
+      id: bonusActivation.id as BonusCode & NonEmptyString,
+      kind: "INewBonusActivation",
+      status: BonusActivationStatusEnum.FAILED
+    };
+    return bonusActivationModel.createOrUpdate(
+      bonusToUpdate,
+      bonusActivation.id
+    );
+  });
+};
+
+const deleteEligibilityCheck = (
+  eligibilityCheckModel: EligibilityCheckModel,
+  bonusActivation: BonusActivationWithFamilyUID
+): TaskEither<Error, string> => {
+  return fromQueryEither(() =>
+    eligibilityCheckModel.deleteOneById(
+      bonusActivation.id as BonusCode & NonEmptyString
+    )
+  );
+};
+
+/**
+ * Operations to be perfomed in case the bonus request was rejected by ADE
+ *
+ *
+ * @returns either a success or a failure request
+ * @throws when the response is considered a transient failure and thus is not considered a domain message
+ */
+export function FailedBonusActivationHandler(
+  bonusActivationModel: BonusActivationModel,
+  bonusLeaseModel: BonusLeaseModel,
+  eligibilityCheckModel: EligibilityCheckModel
+): IFailedBonusActivationHandler {
+  return async (
+    context: Context,
+    input: unknown
+  ): Promise<FailedBonusActivationResult> => {
+    return taskEither
+      .of<FailedBonusActivationFailure, void>(void 0)
+      .chain(_ =>
+        fromEither(FailedBonusActivationInput.decode(input))
+          .mapLeft(() => InvalidInputFailure.encode({ kind: "INVALID_INPUT" }))
+          .map(({ bonusActivation }) => bonusActivation)
+      )
+      .chain(bonusActivation =>
+        deleteEligibilityCheck(eligibilityCheckModel, bonusActivation)
+          // just ignore this error
+          .foldTaskEither(
+            err => {
+              context.log.warn(
+                `FailedBonusActivationHandler|WARN|Failed deleting dsu: ${err.message}`
+              );
+              return taskEither.of(bonusActivation);
+            },
+            _ => taskEither.of(bonusActivation)
+          )
+      )
+      .chain(bonusActivation =>
+        updateBonusAsFailed(bonusActivationModel, bonusActivation)
+          // just ignore this error
+          .foldTaskEither(
+            err => {
+              context.log.warn(
+                `FailedBonusActivationHandler|WARN|Failed updating bonus: ${err.message}`
+              );
+              return taskEither.of(bonusActivation);
+            },
+            _ => taskEither.of(bonusActivation)
+          )
+      )
+      .chain(bonusActivation =>
+        relaseLockForUserFamily(
```

Comment:
I see only now that we have a `UnlockBonusActivationActivity` in the orchestrator which is called regardless of the result. That activity must be removed from the orchestrator since we are going to relase the lock only when thie bonus activation fails here

Reply:
Should we use this activity, and let the orchestrator arrange the flow, or we should do the unlock inside other activities?

Reply:
Ok, I figured it out: I think we should let the orchestrator do the job using a dedicated activity, so we can use native retry mechanism

Reply:
Yes that's probably better since in that way you avoid to call deletedsu and updatebonus more than once if they succeed